text/3470-crabi-v1.md

```diff
@@ -0,0 +1,661 @@
+indirectly via other crABI-supported types).
+
+crABI does not extend the target C ABI in any ways that would require the use
+of target-specific assembly language to support, such as the use of CPU flags
+or registers not already defined by the C ABI. Doing so would require specific
+enablement work in other languages or runtimes, rather than just building on
+existing support for C FFI.
+
+# Rust language interface and interaction with other features
+[language-interface]: #language-interface
+
+The Rust implementation of crABI will include:
+- An ABI for defining functions or referencing external functions:
+  `extern "crabi"`.
+- A repr for laying out data structures (`struct`, `union`, `enum`) compatible
+  with crABI: `repr(crabi)`.
+- A new lint `improper_crabi_types`, analogous to `improper_ctypes`. This lint
+  defaults to `deny`.
+
+crABI structures and functions also support any structure defined using
+`repr(C)`, as well as enums that have a discriminant type specified.
+
+crABI structures and functions support types defined using `repr(transparent)`
+if they would support the type of the field.
+
+For an enum, the crABI `repr` allows additionally specifying the discriminant
+type: `repr(crabi, u8)`. If an `enum` specifies `repr(crabi)` but does not
+specify a discriminant type, the enum is guaranteed to use the smallest
+discriminant type that holds the maximum discriminant value used by a variant
+in the enum. (This differs from the behavior of `repr(C)` enums without a
+discriminant type.)
+
+crABI supports passing and storing pointers to `repr(Rust)` structures, to
+allow using these opaque pointers as handles, but does not support passing such
+structures by value.
+
+crABI works in combination with `repr(align)`, with the same meaning.
+
+crABI does not support `repr(packed)`.
+
+# crABI versioning and evolution
+[crabi-versioning-and-evolution]: #crabi-versioning-and-evolution
+
+crABI has has a major and minor version number, similar to semver.
+
+The major version number will change if crABI changes *incompatibly*; for
+instance, crABI 2.0 would not necessarily interoperate with crABI 1.3.
+
+The minor version number will change if crABI changes *compatibly*, such that
+newer callers can understand and work with older functions and structures.
+Older callers cannot necessarily understand and work with newer functions and
+structures, unless the newer side restricts itself to features understood by an
+older version.
+
+This RFC defines crABI 1.0.
+
+Versions of crABI will start out as unstable features, and follow the usual
+Rust stabilization process. As with any other nightly feature, versions of
+crABI may, prior to stabilization, change in ways incompatible with prior
+*nightly-only* implementations of that version of crABI. Versions of crABI
+should not be considered stable until available in stable Rust.
+
+Future versions of crABI may also establish allow-by-default lints for the use
```

Comment:
As a (wildly excited) outsider, this raises a question for me: will the deny-activation of such lints be enough info for the compiler to restrict itself to a lower subset of crABI?

Or must we reserve some kind of "version" specifier in the `repr` attribute? Something like `repr(crABI, u8, v=2.1)`. perhaps a global annotation on the module level, to avoid repetition?

I'm thinking of cases like an enum that would have different possibilities of niche-optimisation under crABI v1, v2, v3..
Does the compiler have access to the active lint selection in the enum layout code?
Are lints even supposed to act like compiler flags to this extent?